x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/PrefixInputStream.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.repositories.encrypted;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Objects;
+
+/**
+ * A {@code PrefixInputStream} wraps another input stream and exposes
+ * only the first bytes of it. Reading from the wrapping
+ * {@code PrefixInputStream} consumes the underlying stream. The stream
+ * is exhausted when {@code length} bytes have been read or the underlying
+ * stream is exhausted.
+ * <p>
+ * If the {@code closeSource} constructor argument is {@code true}, closing this
+ * stream will also close the underlying input stream. Any subsequent {@code read},
+ * {@code skip} and {@code available} calls will throw {@code IOException}s.
+ */
+public final class PrefixInputStream extends FilterInputStream {
```

Comment:
I also don't like the `FilterInputStream` but figured that since it's there and JDK classes use it, I might use it as well.
I will make all streams, not only the `PrefixInputStream`, extend `InputStream` instead of `FilterInputStream`.